R/combine.R

```diff
@@ -0,0 +1,196 @@
+model_combine_domain <- function(a, b, parameters) {
+  domain <- matrix(NA_real_, length(parameters), 2)
+  rownames(domain) <- parameters
+
+  a_only <- setdiff(a$parameters, b$parameters)
+  domain[a_only, ] <- a$domain[a_only, ]
+
+  b_only <- setdiff(b$parameters, a$parameters)
+  domain[b_only, ] <- b$domain[b_only, ]
+
+  ## Shared parameters, take the most restrictive:
+  shared <- intersect(a$parameters, b$parameters)
+  domain[shared, 1] <- pmax(a$domain[shared, 1], b$domain[shared, 1])
+  domain[shared, 2] <- pmin(a$domain[shared, 2], b$domain[shared, 2])
+
+  domain
+}
+
+
+model_combine_density <- function(a, b, parameters) {
+  i_a <- match(a$parameters, parameters)
+  i_b <- match(b$parameters, parameters)
+  function(x, ...) {
+    a$density(x[i_a], ...) + b$density(x[i_b], ...)
+  }
+}
+
+
+model_combine_gradient <- function(a, b, parameters, properties, call = NULL) {
+  if (isFALSE(properties$has_gradient)) {
+    return(NULL)
+  }
+  possible <- a$properties$has_gradient && b$properties$has_gradient
+  required <- isTRUE(properties$has_gradient)
+  if (!possible && !required) {
+    return(NULL)
+  }
+  if (required && !possible) {
+    cli::cli_abort(
+      c("Can't create a gradient from these models",
+        i = paste("Both models must have a gradient in order to be able",
+                  "to create a gradient from the combination")),
+      call = call)
+  }
+
+  n_pars <- length(parameters)
+  i_a <- match(a$parameters, parameters)
+  i_b <- match(b$parameters, parameters)
+  function(x, ...) {
+    ret <- numeric(n_pars)
+    ret[i_a] <- ret[i_a] + a$gradient(x[i_a], ...)
+    ret[i_b] <- ret[i_b] + b$gradient(x[i_b], ...)
+    ret
+  }
+}
+
+
+model_combine_direct_sample <- function(a, b, parameters, properties,
+                                        call = NULL) {
+  if (isFALSE(properties$has_direct_sample)) {
+    return(NULL)
+  }
+  possible <- a$properties$has_direct_sample != b$properties$has_direct_sample
+  required <- isTRUE(properties$has_direct_sample)
+  if (!possible && !required) {
+    return(NULL)
+  }
+  if (required && !possible) {
+    if (a$properties$has_direct_sample) {
+      hint <- paste("Both models have a 'direct_sample' method so we can't",
+                    "combine them. Set 'has_direct_sample = FALSE' on one",
+                    "of your models and try again")
+    } else {
+      hint <- "Neither of your models have 'direct_sample' methods"
+    }
+    cli::cli_abort(
+      c("Can't create a direct_sample from these models",
+        i = hint),
+      call = call)
+  }
+
+  ## Next we need to check that the model with a sample has the full set of
+  if (a$properties$has_direct_sample) {
+    name <- "a"
+    model <- a
+  } else {
+    name <- "b"
+    model <- b
+  }
+  if (!all(parameters %in% model$parameters)) {
+    cli::cli_abort(
+      paste("Can't create a direct_sample from these models as '{name}' does",
+            "not contain all parameters in both models"),
```

Comment:
im not sure that the name "a" or "b" is useful to the user, i found it actively confusing in the test for this, can we do a
```R
`+.mcstate_model` <- function(x, y,
  names = c(deparse(substitute(x)),deparse(substitute(y))))
```
and pass this down into this function so that we can reference the variable that they tried to sum?